### PR TITLE
Implement efficient size method for Tries

### LINKED ISSUE #13622: Stackoverflow while loading a UI page
**Jetty version(s)**

**Jetty Environment**

jetty 12_0_16
**Java version/vendor** `(use: java -version)`
Java version: 17.0.14,17.0.14+7,IBM Corporation
**OS type/version**
OS-System: Linux 4.18.0-553.56.1.el8_10.x86_64,amd64
**Description**

We are using smartclient 12.1 version for deploying our webapp. We recently upgraded to this version and seeing stackoverflow when the UI loads.

ERROR LoadDSTag - Exception while attempting to process a loadDS tag.
java.lang.StackOverflowError: null
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:498) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:494) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.keySet(TreeTrie.java:480) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.TreeTrie.size(TreeTrie.java:309) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.ClassMatcher$ByPackage.size(ClassMatcher.java:181) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.ClassMatcher$ByPackageOrName.size(ClassMatcher.java:284) ~[jetty-util-12.0.16.jar:12.0.16]
at java.util.AbstractCollection.isEmpty(AbstractCollection.java:89) ~[?:?]
at org.eclipse.jetty.util.IncludeExcludeSet.isIncludedAndNotExcluded(IncludeExcludeSet.java:204) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.ClassMatcher.combine(ClassMatcher.java:810) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.util.ClassMatcher.match(ClassMatcher.java:759) ~[jetty-util-12.0.16.jar:12.0.16]
at org.eclipse.jetty.ee8.webapp.WebAppContext.isServerClass(WebAppContext.java:705) ~[jetty-ee8-webapp-12.0.16.jar:12.0.16]
at org.eclipse.jetty.ee8.webapp.WebAppContext.isHiddenClass(WebAppContext.java:695) ~[jetty-ee8-webapp-12.0.16.jar:12.0.16]
at org.eclipse.jetty.ee8.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:393) ~[jetty-ee8-webapp-12.0.16.jar:12.0.16]
at java.lang.ClassLoader.loadClass(ClassLoader.java:1095) ~[?:?]
at org.apache.xerces.parsers.ObjectFactory.findProviderClass(Unknown Source) ~[xercesImpl.jar:?]
at org.apache.xerces.parsers.ObjectFactory.newInstance(Unknown Source) ~[xercesImpl.jar:?]
at org.apache.xerces.parsers.ObjectFactory.createObject(Unknown Source) ~[xercesImpl.jar:?]
at org.apache.xerces.parsers.ObjectFactory.createObject(Unknown Source) ~[xercesImpl.jar:?]
at org.apache.xerces.parsers.DOMParser.<init>(Unknown Source) ~[xercesImpl.jar:?]
at org.apache.xerces.parsers.DOMParser.<init>(Unknown Source) ~[xercesImpl.jar:?]
at org.apache.xerces.jaxp.DocumentBuilderImpl.<init>(Unknown Source) ~[xercesImpl.jar:?]
at org.apache.xerces.jaxp.DocumentBuilderFactoryImpl.newDocumentBuilder(Unknown Source) ~[xercesImpl.jar:?]
at com.isomorphic.xml.XML.parseXML(XML.java:295) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.xml.XML.parseXML(XML.java:152) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.xml.XML.toDSRecords(XML.java:498) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.xml.XML$RecordsFromXMLCache.loadObjectFromFile(XML.java:664) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.store.ProcessedFileCache.getObjectFromFile(ProcessedFileCache.java:147) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.xml.XML.loadCacheableDSRecords(XML.java:676) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.store.DataStructCache.loadInstance(DataStructCache.java:187) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.datasource.DataSource.loadDS(DataSource.java:569) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.datasource.DataSource.forName(DataSource.java:544) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.datasource.PoolableDataSourceFactory.makeUnpooledObject(PoolableDataSourceFactory.java:125) ~[isomorphic_core_rpc.jar:?]
at com.isomorphic.datasource.PoolableDataSourceFactory.makeObject(PoolableDataSourceFactory.java:150) ~[isomorphic_core_rpc.jar:?]
at org.apache.commons.pool.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:1220) ~[commons-pool-1.6.jar:1.6]
at com.isomorphic.pool.PoolManager.borrowObject(PoolManager.java:85) ~[isomorphic_core_rpc.jar:?]
at

**How to reproduce?**

This issue is occuring only after few days of usage. And once it occurs the page never loads again. The issue was not seen with jetty 9_4_57. Its only seen after upgrading to the version mentioned.

### PR DESCRIPTION
**Pull request by gregw**
The isEmpty method can be frequently called on tries and the current implementation creates a disposible key set each time.

### CODE REVIEW

**Comment by gregw** [comment]
Seen in #13622 - whilst not the cause, the depth was not good either.
It may also be worthwhile looking to avoid tail recursion when possible.,

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java [hunk 1]
```diff
@@ -137,6 +138,7 @@ public String toString()
      *
      * The array is of characters rather than integers to save space.
      */
+    private final AtomicInteger _size = new AtomicInteger();
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java]: `ArrayTrie` isn't thread-safe, so that atomic can be replaced by a simple int.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java [hunk 2]
```diff
@@ -97,6 +98,7 @@ private Node(char c)
         }
     }
 
+    private final AtomicInteger _size = new AtomicInteger();
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java]: `TreeTrie` also isn't thread-safe, so int can be used here too.

**Review by lorban — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java [hunk 3]
```diff
@@ -296,6 +299,10 @@ else if (idx >= big.length)
         if (node == null)
             node = _node[row] = new Node<>();
         node._key = key;
+        if (node._value == null && value != null)
+            _size.incrementAndGet();
+        else if (node._value != null && value == null)
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java:303]: The javadocs of the `put()` method say that `value` cannot be `null`, but `remove()` obviously breaks that.

Furthermore, what's the point of creating the entry, just to have a `null` value at the end?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java [hunk 4]
```diff
@@ -156,8 +159,20 @@ public boolean put(String s, V v)
                 t = n;
             }
         }
-        t._key = v == null ? null : s;
-        t._value = v;
+        if (v == null)
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java:161]: Same comment... This is creating all the nodes.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java (hunk 1 above)
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java]: Why atomic? It's supposed to be non-concurrent.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java (hunk 3 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java:303]: This PR is just adding size.  I don't want to change the way the Trie works.  If a node was created before it is still created after this PR

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java (hunk 4 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java:161]: same response.  This PR is not changing the way the Trie works, just adding a size

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java (hunk 3 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java:303]: Also we actually have tests that check we can put null values, so I have changed the javadoc

File: jetty-core/jetty-server/src/test/resources/jetty-logging.properties [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#org.eclipse.jetty.LEVEL=DEBUG
+Doorg.eclipse.jetty.LEVEL=DEBUG
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/resources/jetty-logging.properties]: Please revert.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java [hunk 6]
```diff
@@ -276,37 +290,13 @@ private V getBest(Node<V> node, byte[] b, int offset, int len)
...
-                }
-            }
-            for (int i = t._nextOther.size(); i-- > 0; )
-            {
-                if (!isEmpty(t._nextOther.get(i)))
-                    return false;
-            }
-        }
-        return true;
+        return _size == 0;
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java]:
```suggestion
        return size() == 0;
```

**Review by sbordet — CHANGES_REQUESTED** [review]

**Review by lorban — APPROVED** [review]

**Review by sbordet — APPROVED** [review]
